Source/AwsCommonRuntimeKit/crt/AWSString.swift

```diff
@@ -3,38 +3,31 @@
 import AwsCCommon
 
 final class AWSString {
-    let string: String
     let rawValue: UnsafeMutablePointer<aws_string>
 
     init(_ str: String, allocator: Allocator) {
-        self.string = str
         self.rawValue = aws_string_new_from_array(allocator.rawValue, str, str.count)
     }
 
     var count: Int {
         return self.rawValue.pointee.len
     }
 
-    func newByteCursor() -> ByteCursor {
-        return AWSStringByteCursor(self)
-    }
-
-    func asCStr() -> UnsafePointer<Int8> {
-        return aws_string_c_str(self.rawValue)
-    }
-
     deinit {
         aws_string_destroy(self.rawValue)
     }
 }
 
-private struct AWSStringByteCursor: ByteCursor {
-    private let awsString: AWSString
-    public var rawValue: aws_byte_cursor
+// TODO: try to find a better way to get a long lived byte cursor from String without making a copy.
+class AWSStringByteCursor {
+    let awsString: AWSString
+    let string: String
+    let byteCursor: aws_byte_cursor
 
-    init(_ awsString: AWSString) {
-        self.awsString = awsString
-        self.rawValue = aws_byte_cursor_from_string(awsString.rawValue)
+    init(_ string: String, allocator: Allocator = defaultAllocator) {
```

Comment:
for internal classes, we SHOULDN'T provide a default allocator.
if some public class needs to use an internal class, it should be forced to pass along its allocator

If addressing this is a huge revamp, just mark it with TODO

Personally, I'm not sure it's worth supporting per-object custom allocators AT ALL for aws-crt-swift. It's just not a thing that's done in the Swift ecosystem.

Reply:
Makes sense, thanks.

Reply:
Thanks, added a todo. I guess it’s nice to have it for the tests where we it can help detect memory leaks. We can also maybe just override the defaultAllocator with tracing allocator for tests.